File: Leetcode/020_ValidParentheses.cc
#include<string>
#include<stack>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char>parentheses;
        for(int i=0;i<s.size();i++){
            if(s[i]=='('||s[i]=='{'||s[i]=='['){
                parentheses.push(s[i]);
            }
            else{
                if(parentheses.empty())
                    return false;
                if(parentheses.top()=='('&&s[i]==')'){
                    parentheses.pop();
                    continue;
                }
                if(parentheses.top()=='{'&&s[i]=='}'){
                    parentheses.pop();
                    continue;
                }
                if(parentheses.top()=='['&&s[i]==']'){
                    parentheses.pop();
                    continue;
                }
                return false;
            }
        }
        if(!parentheses.empty())return false;
        return true;
    }
};

File: Leetcode(python)/020_ValidParentheses.py
# coding:utf-8
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        #使用列表作为栈
        stack = []
        dict = {"]":"[", "}":"{", ")":"("}
        for char in s:
            if char in dict.values(): #这样可以判断values
                stack.append(char)
            elif char in dict.keys(): #这样可以判断keys
                if stack == [] or dict[char] != stack.pop(): #列表做栈
                    return False
            else:
                return False
        return stack == []